auto_round/smooth_quant/graph_trace.py

```diff
@@ -0,0 +1,222 @@
+
+    def skip_op_absorb_helper(self, parent_node):
+        for val_user in list(parent_node.outputs())[0].uses():
+            next_node = val_user.user
+            if next_node.kind() == "aten::size":
+                continue
+            elif next_node.kind() in self.could_absorb_layers:
+                continue
+            elif next_node.kind() in self.skip_ops_to_find_absorb:
+                node_res = self.skip_op_absorb_helper(next_node)
+                if not node_res:
+                    return False
+            else:
+                return False
+        return True
+
+    def mapping_torch_module_to_aten(self, op_types):
+        res = []
+        for op in op_types:
+            if op not in self.supported_torch_module_to_aten.keys():
+                logger.warning(f"{op} is not supported in smooth quant, ignoring...")
+                continue
+            res.append(self.supported_torch_module_to_aten[op])
+        res = list(set(res))
+        return res
+
+    def _check_valid_conv(self, module):
+        """Remove group conv except depthwise conv
+        :param module:
+
+        :return:
+        """
+        if not isinstance(module, torch.nn.Conv2d):
+            return True
+        if module.groups > 1:
+            if module.in_channels == module.out_channels and module.groups == module.in_channels:
+                return True
+            else:
+                return False
+        return True
+
+    def get_absorb_to_layer(self, model, example_input, op_types, skip_unsupported_layers=True):
+        traced_model = self.trace(model, example_input)
+        if traced_model is None:
+            return None, None
+
+        aten_op_types = self.mapping_torch_module_to_aten(op_types)
+        nodes_types = self.get_nodes(traced_model, aten_op_types)
+        nodes = [node_type[0] for node_type in nodes_types]
+        nodes_prev_absorb = self.get_prev_absorb_layer(nodes)
+        absorb_to_layer = {}
+        no_absorb_layers = []
+        for index, absorb in enumerate(nodes_prev_absorb):
+            if absorb is None:
+                no_absorb_layers.append(".".join(nodes[index].scopeName().split("/")[-1].split(".")[1:]))
+                continue
+            node = nodes[index]
+            layer_name = ".".join(node.scopeName().split("/")[-1].split(".")[1:])
+            absorb_name = ".".join(absorb.scopeName().split("/")[-1].split(".")[1:])
+            if layer_name == "" or absorb_name == "":
+                continue
+            if absorb_name in absorb_to_layer.keys():
+                absorb_to_layer[absorb_name].append(layer_name)
+            else:
+                absorb_to_layer[absorb_name] = [layer_name]
+        if skip_unsupported_layers:
+            absorb_to_layer = self.remove_unsupported_layers(model, absorb_to_layer, no_absorb_layers)
+        return absorb_to_layer, no_absorb_layers
+
+    def remove_unsupported_layers(self, model, absorb_to_layer, no_absorb_layers):
```

Comment:
refer to autoawq to support handcrafted fused patten for popular models and use these configs first. graph trace has some limitations